design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+---
+
+It would also be somewhat helpful in case of canary deployments when there's
```

Comment:
Hm, I'm not fully aware of solutions like flagger and how they handle canary deployments. 

> For canary, wouldn't you just shift the load back

But yes, general idea would be shifting back the load. In case, there's circuit breaker configured, new instances would be circuit break-ed and traffic would be sent only to old instances i.e. in a way we will be shifting the load back. Though yeah, I'm not fully sure how it would fit well with other available tools in a bigger picture.